internal/controllers/configuration/kongadminapi_controller.go

```diff
@@ -0,0 +1,102 @@
+package configuration
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/samber/lo"
+	corev1 "k8s.io/api/core/v1"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/adminapi"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/util"
+)
+
+// KongAdminAPIServiceReconciler reconciles Kong Admin API Service Endpointslices
+// and notifies the provided notifier about those.
+type KongAdminAPIServiceReconciler struct {
+	client.Client
+
+	ServiceNN types.NamespacedName
+	Service   *corev1.Service
+
+	Log              logr.Logger
+	CacheSyncTimeout time.Duration
+
+	EndpointsNotifier EndpointsNotifier
+}
+
+type EndpointsNotifier interface {
+	Notify(addresses []string)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *KongAdminAPIServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	c, err := controller.New("KongAdminAPIEndpoints", mgr, controller.Options{
+		Reconciler: r,
+		LogConstructor: func(_ *reconcile.Request) logr.Logger {
+			return r.Log
+		},
+		CacheSyncTimeout: r.CacheSyncTimeout,
+	})
+	if err != nil {
+		return err
+	}
+
+	return c.Watch(
+		&source.Kind{Type: &discoveryv1.EndpointSlice{}},
+		&handler.EnqueueRequestForObject{},
+		predicate.NewPredicateFuncs(r.shouldReconcileEndpointSlice),
+	)
+}
+
+func (r *KongAdminAPIServiceReconciler) shouldReconcileEndpointSlice(obj client.Object) bool {
+	endpoints, ok := obj.(*discoveryv1.EndpointSlice)
+	if !ok {
+		return false
+	}
+
+	if !lo.ContainsBy(endpoints.OwnerReferences, func(ref metav1.OwnerReference) bool {
+		return ref.Kind == "Service" && ref.Name == r.ServiceNN.Name
```

Comment:
Shouldn't we check that the namespace is the same as well?

Reply:
`metav1.OwnerReference` doesn't have a `Namespace` field. We can check

```
	if endpoints.Namespace != r.ServiceNN.Namespace {
		return false
	}
```

Reply:
👍

Reply:
Added